File: mobile/components/PasswordRules.js
import { View, Text, StyleSheet } from 'react-native';
import { useTheme } from '../theme';

// The backend requires 8+ characters. The rest are suggestions.
export const MIN_PASSWORD_LENGTH = 8;

export function checkPassword(password) {
  return {
    length: password.length >= MIN_PASSWORD_LENGTH,
    upper: /[A-Z]/.test(password),
    number: /[0-9]/.test(password),
    symbol: /[^A-Za-z0-9]/.test(password),
  };
}

// Only length rule blocks submission.
export function isPasswordValid(password) {
  return checkPassword(password).length;
}

function Rule({ met, label, required }) {
  const { colors } = useTheme();
  return (
    <View style={styles.row}>
      <Text
        style={{
          fontSize: 13,
          width: 18,
          color: met ? colors.success : required ? colors.error : colors.textMuted,
        }}
      >
        {met ? '✓' : '✗'}
      </Text>
      <Text
        style={{
          fontSize: 13,
          color: met ? colors.text : colors.textMuted,
          flex: 1,
        }}
      >
        {label}
        {required ? '' : '  (recommended)'}
      </Text>
    </View>
  );
}

export default function PasswordRules({ password }) {
  const { colors } = useTheme();
  if (!password) return null;

  const c = checkPassword(password);
  const strength = [c.length, c.upper, c.number, c.symbol].filter(Boolean).length;

  return (
    <View
      style={[
        styles.box,
        { backgroundColor: colors.surfaceAlt, borderColor: colors.border },
      ]}
    >
      <View style={styles.bars}>
        {[0, 1, 2, 3].map((i) => (
          <View
            key={i}
            style={[
              styles.bar,
              {
                backgroundColor:
                  i < strength
                    ? strength <= 1
                      ? colors.error
                      : strength <= 2
                        ? colors.warning
                        : colors.success
                    : colors.border,
              },
            ]}
          />
        ))}
      </View>

      <Rule met={c.length} required label={`At least ${MIN_PASSWORD_LENGTH} characters`} />
      <Rule met={c.upper} label="An uppercase letter" />
      <Rule met={c.number} label="A number" />
      <Rule met={c.symbol} label="A symbol" />
    </View>
  );
}

const styles = StyleSheet.create({
  box: {
    borderWidth: 1,
    borderRadius: 10,
    padding: 12,
    marginTop: 10,
    gap: 4,
  },
  bars: { flexDirection: 'row', gap: 5, marginBottom: 10 },
  bar: { flex: 1, height: 4, borderRadius: 2 },
  row: { flexDirection: 'row', alignItems: 'center', gap: 6 },
});

File: noteriety_mobile/lib/widgets/password_rules.dart
import 'package:flutter/material.dart';

import '../theme/app_colors.dart';

const int minPasswordLength = 8;

class PasswordChecks {
  final bool length;
  final bool upper;
  final bool number;
  final bool symbol;
  const PasswordChecks(this.length, this.upper, this.number, this.symbol);

  int get strength =>
      [length, upper, number, symbol].where((b) => b).length;
}

PasswordChecks checkPassword(String password) {
  return PasswordChecks(
    password.length >= minPasswordLength,
    RegExp(r'[A-Z]').hasMatch(password),
    RegExp(r'[0-9]').hasMatch(password),
    RegExp(r'[^A-Za-z0-9]').hasMatch(password),
  );
}

bool isPasswordValid(String password) => checkPassword(password).length;

class PasswordRules extends StatelessWidget {
  final String password;
  const PasswordRules({super.key, required this.password});

  @override
  Widget build(BuildContext context) {
    if (password.isEmpty) return const SizedBox.shrink();
    final colors = context.colors;
    final c = checkPassword(password);
    final strength = c.strength;

    Color barColor(int i) {
      if (i >= strength) return colors.border;
      if (strength <= 1) return colors.error;
      if (strength <= 2) return colors.warning;
      return colors.success;
    }

    return Container(
      margin: const EdgeInsets.only(top: 10),
      padding: const EdgeInsets.all(12),
      decoration: BoxDecoration(
        color: colors.surfaceAlt,
        border: Border.all(color: colors.border),
        borderRadius: BorderRadius.circular(10),
      ),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Padding(
            padding: const EdgeInsets.only(bottom: 10),
            child: Row(
              children: List.generate(4, (i) {
                return Expanded(
                  child: Container(
                    height: 4,
                    margin: EdgeInsets.only(right: i < 3 ? 5 : 0),
                    decoration: BoxDecoration(
                      color: barColor(i),
                      borderRadius: BorderRadius.circular(2),
                    ),
                  ),
                );
              }),
            ),
          ),
          _Rule(
            met: c.length,
            required: true,
            label: 'At least $minPasswordLength characters',
          ),
          _Rule(met: c.upper, label: 'An uppercase letter'),
          _Rule(met: c.number, label: 'A number'),
          _Rule(met: c.symbol, label: 'A symbol'),
        ],
      ),
    );
  }
}

class _Rule extends StatelessWidget {
  final bool met;
  final bool required;
  final String label;
  const _Rule({required this.met, required this.label, this.required = false});

  @override
  Widget build(BuildContext context) {
    final colors = context.colors;
    final markColor =
        met ? colors.success : (required ? colors.error : colors.textMuted);
    return Padding(
      padding: const EdgeInsets.symmetric(vertical: 2),
      child: Row(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          SizedBox(
            width: 18,
            child: Text(
              met ? '\u2713' : '\u2717',
              style: TextStyle(fontSize: 13, color: markColor),
            ),
          ),
          Expanded(
            child: Text(
              required ? label : '$label  (recommended)',
              style: TextStyle(
                fontSize: 13,
                color: met ? colors.text : colors.textMuted,
              ),
            ),
          ),
        ],
      ),
    );
  }
}
